File: path-sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// dfs: if there is a sum, shore up that sum alone. 
// in main func, return true if helper val is the sum
// backtrack of course to correct the sum
//         5
//       /   \
//      4     8
// 13 -- true: 
// ts: 8
// 4 -- false
func hasPathSum(root *TreeNode, targetSum int) bool {
    if root == nil { return false }
    targetSum -= root.Val
    if root.Left == nil && root.Right == nil {
        return targetSum == 0
    }
    return hasPathSum(root.Left, targetSum) || hasPathSum(root.Right, targetSum)
}

File: path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
#       5  9
#     /    
#    4      
# q = [(5, 13)-pop- (4,13-5), (8, 13-5)]
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root: 
            return False
        q = deque([(root, targetSum)])
        while q:
            root, target = q.popleft()
            if not root.left and not root.right and root.val == target:
                return True
            if root.left:
                q.append((root.left, target - root.val))
            if root.right:
                q.append((root.right, target - root.val))
        return False

